crates/red_knot_python_semantic/src/suppression.rs

```diff
@@ -102,3 +158,478 @@ pub(crate) struct Suppression {
+        let cursor = Cursor::new(&source[range]);
+
+        Self { cursor, range }
+    }
+
+    fn parse_comment(&mut self) -> Option<SuppressionComment> {
+        let comment_start = self.offset();
+        self.cursor.start_token();
+
+        if !self.cursor.eat_char('#') {
+            return None;
+        }
+
+        self.eat_whitespace();
+
+        // type: ignore[code]
+        // ^^^^^^^^^^^^
+        let kind = self.eat_kind()?;
+
+        let has_trailing_whitespace = self.eat_whitespace();
+
+        // type: ignore[code1, code2]
+        //             ^^^^^^
+        let codes = self.eat_codes();
+
+        if self.cursor.is_eof() || codes.is_some() || has_trailing_whitespace {
+            // Consume the comment until its end or until the next "sub-comment" starts.
+            self.cursor.eat_while(|c| c != '#');
+            Some(SuppressionComment {
+                kind,
+                codes,
+                range: TextRange::at(comment_start, self.cursor.token_len()),
+            })
+        } else {
+            None
+        }
+    }
+
+    fn eat_kind(&mut self) -> Option<SuppressionKind> {
+        let kind = if self.cursor.as_str().starts_with("type") {
+            SuppressionKind::TypeIgnore
+        } else if self.cursor.as_str().starts_with("knot") {
+            SuppressionKind::Knot
+        } else {
+            return None;
+        };
+
+        self.cursor.skip_bytes(kind.len_utf8());
+
+        self.eat_whitespace();
+
+        if !self.cursor.eat_char(':') {
+            return None;
+        }
+
+        self.eat_whitespace();
+
+        if !self.cursor.as_str().starts_with("ignore") {
+            return None;
+        }
+
+        self.cursor.skip_bytes("ignore".len());
+
+        Some(kind)
+    }
+
+    fn eat_codes(&mut self) -> Option<SmallVec<[TextRange; 2]>> {
+        if !self.cursor.eat_char('[') {
+            return None;
+        }
+
+        let mut codes: SmallVec<[TextRange; 2]> = smallvec![];
+
+        loop {
+            if self.cursor.is_eof() {
+                return None;
+            }
+
+            self.eat_whitespace();
+
+            // `knot: ignore[]` or `knot: ignore[a,]`
+            if self.cursor.eat_char(']') {
+                break Some(codes);
+            }
+
+            let code_start = self.offset();
+            if !self.eat_word() {
+                return None;
+            }
+
+            codes.push(TextRange::new(code_start, self.offset()));
+
+            self.eat_whitespace();
+
+            if !self.cursor.eat_char(',') {
+                self.eat_whitespace();
+
+                if self.cursor.eat_char(']') {
+                    break Some(codes);
+                }
+                // `knot: ignore[a b]
+                return None;
+            }
+        }
+    }
+
+    fn eat_whitespace(&mut self) -> bool {
+        if self.cursor.eat_if(char::is_whitespace) {
+            self.cursor.eat_while(char::is_whitespace);
```

Comment:
Should we use `is_python_whitespace`?

Reply:
Using `is_whitespace` should be fine because we're in a comment. This also matches the formatter's docstring handling